### PR TITLE
expression: return null when cast to huge binary type

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Before this PR, for this query:
```
mysql> select cast(1 as binary(4294967295));
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
and tidb-server is down, because we would allocate about 4GB memory for the result, so it triggers golang runtime panic "out of memory" on my machine, but we don't catch this panic and do recover.

### What is changed and how it works?

According to https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast, MySQL should also allocate this 4GB memory for binary result, and pad `\0` in the tail, but in my practical experiment, it does not. Here is the output of MySQL 5.7.10:
```
mysql> select cast(1 as binary(4294967295));
+-------------------------------+
| cast(1 as binary(4294967295)) |
+-------------------------------+
| 1                             |
+-------------------------------+
1 row in set (0.00 sec)

mysql> select cast(1 as binary(4294967295)) = '1';
+-------------------------------------+
| cast(1 as binary(4294967295)) = '1' |
+-------------------------------------+
|                                   1 |
+-------------------------------------+
1 row in set (0.00 sec)
```
In other word, MySQL does not behave as its documentation on this issue.

We have 2 choices to fix this problem:
- recover from the "out of memory" panic and report error;
- check the length of target type of cast function, according to https://dev.mysql.com/doc/refman/5.7/en/binary-varbinary.html, binary type has maximum length of 255;

~~these 2 choices may lead to different behaviors, I prefer the second one, because if we impose no check on the length and recover when panic occurs, it has risk to trigger panic in other places since binary type is assumed to be less than 255 bytes.~~

Based on https://github.com/pingcap/tidb/pull/8768#discussion_r243640356, change the behavior to be compatible with MySQL 8.0.

### Check List 

Tests 

 - Unit test

Code changes

N/A

Side effects

 - Breaking backward compatibility: ~~as mentioned above, `select cast(1 as binary(1024))` is able to be executed successfully before, but unable now.~~ No compatibility breaking now.

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8768)

### CODE REVIEW

**Comment by eurekaka** [comment]
@zz-jason @winoros @lamxTyler @XuHuaiyu PTAL

File: planner/core/preprocess_test.go [hunk 1]
```diff
@@ -197,6 +197,7 @@ func (s *testValidatorSuite) TestValidator(c *C) {
 		{"select * from (select 1 ) a , (select 2) b, (select * from (select 3) a join (select 4) b) c;", false, nil},
 
 		{"CREATE VIEW V (a,b,c) AS SELECT 1,1,3;", false, nil},
+		{"select cast(1 as binary(1024));", false, errors.New("[types:1074]Field length too big for target type (max = 255)")},
```
Comment (lamxTyler) [inline planner/core/preprocess_test.go]: Did you try this in mysql 5.7? Seems we get the expected result, not errors.
BTW, in mysql 8.0.
```sql

MySQL [(none)]> select cast(1 as binary(4294967295));
+-------------------------------+
| cast(1 as binary(4294967295)) |
+-------------------------------+
| NULL                          |
+-------------------------------+
1 row in set, 1 warning (0.00 sec)

MySQL [(none)]> show warnings;
+---------+------+--------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                              |
+---------+------+--------------------------------------------------------------------------------------+
| Warning | 1301 | Result of cast_as_binary() was larger than max_allowed_packet (67108864) - truncated |
+---------+------+--------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (morgo) [reply planner/core/preprocess_test.go]: I like the MySQL 8.0 behavior (check if the length is greater than `max_allowed_packet` and then return `NULL` + a warning).  I don't think the 5.7 behavior is intended.

Many [MySQL 8.0 features](https://github.com/pingcap/tidb/issues/7968) are already on our roadmap, so it's also not a stretch to think we may change our targeted compatible version.

**Comment by eurekaka** [comment]
@zz-jason @lamxTyler PTAL

**Pull request merged**